File: Cursojava22/app/src/main/java/com/example/cursojava2/UsuariosFirebase/LoginFire.java
package com.example.cursojava2.UsuariosFirebase;

import android.app.ProgressDialog;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Patterns;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.cursojava2.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.util.regex.Pattern;

public class LoginFire extends AppCompatActivity implements View.OnClickListener {
    private static final Pattern PASSWORD_PATTERN =
            Pattern.compile("^" +
                    "(?=.*[0-9])" +         //at least 1 digit
                    "(?=.*[a-z])" +         //at least 1 lower case letter
                    "(?=.*[A-Z])" +         //at least 1 upper case letter
                    //"(?=.*[a-zA-Z])" +      //any letter
                    //"(?=.*[@#$%^&+=])" +    //at least 1 special character
                    //"(?=\\S+$)" +           //no white spaces
                    ".{8,}" +               //at least 8 characters
                    "$");

    private EditText Name;
    private EditText Password;
    private TextView Info;
    private Button Login;
    private int counter = 5;
    private TextView userRegistration;
    private FirebaseAuth firebaseAuth;
    private ProgressDialog progressDialog;
    private TextView forgotPassword;
    String name,password;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login_fire);

        setupActionBar();

        Name = (EditText)findViewById(R.id.etName);
        Password = (EditText)findViewById(R.id.etPassword);
        Info = (TextView)findViewById(R.id.tvInfo);
        Login = (Button)findViewById(R.id.btnLogin);
        Login.setOnClickListener(this);
        userRegistration = (TextView)findViewById(R.id.tvRegister);
        forgotPassword = (TextView)findViewById(R.id.tvForgotPassword);

        Info.setText("Numero de intentos restantes: 5");

        firebaseAuth = FirebaseAuth.getInstance();
        progressDialog = new ProgressDialog(this);

        FirebaseUser user = firebaseAuth.getCurrentUser();

        if(user != null){
            //finish();

            startActivity(new Intent(LoginFire.this, SecondActivity.class));

        }

        Login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //validacion();
                if(validacion()) {
                    validate(Name.getText().toString(), Password.getText().toString());
                }
            }
        });

        userRegistration.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(LoginFire.this, RegAct.class));
            }
        });

        forgotPassword.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(LoginFire.this, PasswordActivity.class));
            }
        });
    }

    private void validate(String userName, String userPassword) {

        progressDialog.setMessage("Espere un momento por favor");
        progressDialog.show();

        firebaseAuth.signInWithEmailAndPassword(userName, userPassword).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if(task.isSuccessful()){
                    progressDialog.dismiss();
                    //Toast.makeText(MainActivity.this, "Login Successful", Toast.LENGTH_SHORT).show();
                    checkEmailVerification();
                }else{
                    Toast.makeText(LoginFire.this, "Verifique que los datos estén correctos ", Toast.LENGTH_SHORT).show();
                    counter--;
                    Info.setText("Numero de intentos restantes: " + counter);
                    progressDialog.dismiss();
                    if(counter == 0){
                        Login.setEnabled(false);
                    }
                }
            }
        });


    }

    private void checkEmailVerification(){
        FirebaseUser firebaseUser = firebaseAuth.getInstance().getCurrentUser();
        Boolean emailflag = firebaseUser.isEmailVerified();

        startActivity(new Intent(LoginFire.this, SecondActivity.class));

    }

    private Boolean validacion(){

        Boolean result = false;

        password = Password.getText().toString();
        name = Name.getText().toString();


        if (name.equals("")){
            Name.setError("Requerido");
        }
        else if (!Patterns.EMAIL_ADDRESS.matcher(name).matches()) {
            Name.setError("Por favor, introduce una dirección de correo electrónico válida");
            return false;
        }
        else if (password.equals("")){
            Password.setError("Requerido");
        }else if (!PASSWORD_PATTERN.matcher(password).matches()) {
            Password.setError("La contraseña no sigue los parametros");
            return false;
        } else {
            Password.setError(null);
            return true;
        }

        return result;
    }

    private boolean validatePassword() {

        if (password.isEmpty()) {
            Password.setError("El campo no puede estar vacío");
            return false;
        } else if (!PASSWORD_PATTERN.matcher(password).matches()) {
            Password.setError("Password too weak");
            return false;
        } else {
            Password.setError(null);
            return true;
        }
    }

    @Override
    public void onClick(View v) {

    }

    private boolean validateEmail() {

        if (name.isEmpty()) {
            Name.setError("El campo no puede estar vacío");
            return false;
        } else if (!Patterns.EMAIL_ADDRESS.matcher(name).matches()) {
            Name.setError("Por favor, introduce una dirección de correo electrónico válida");
            return false;
        } else {
            Name.setError(null);
            return true;
        }
    }



    private void setupActionBar(){
        ActionBar actionBar = getSupportActionBar();
        if(actionBar != null){
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle("Login a Firebase");
        }
    }

}

File: KotlinCurso2/app/src/main/java/com/example/kotlincurso2/UsuariosFirebase/LoginFire.kt
package com.example.kotlincurso2.UsuariosFirebase

import android.app.ProgressDialog
import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Patterns
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import com.example.kotlincurso2.R
import com.google.firebase.auth.FirebaseAuth
import java.util.regex.Pattern

class LoginFire : AppCompatActivity() , View.OnClickListener {

    private val PASSWORD_PATTERN = Pattern.compile(
        "^" +
                "(?=.*[0-9])" +         //at least 1 digit

                "(?=.*[a-z])" +         //at least 1 lower case letter

                "(?=.*[A-Z])" +         //at least 1 upper case letter

                //"(?=.*[a-zA-Z])" +      //any letter
                //"(?=.*[@#$%^&+=])" +    //at least 1 special character
                //"(?=\\S+$)" +           //no white spaces
                ".{8,}" +               //at least 8 characters

                "$"
    )

    private var Name: EditText? = null
    private var Password: EditText? = null
    private var Info: TextView? = null
    private var Login: Button? = null
    private var counter = 5
    private var userRegistration: TextView? = null
    private var firebaseAuth: FirebaseAuth? = null
    private var progressDialog: ProgressDialog? = null
    private var forgotPassword: TextView? = null
    internal lateinit var name:String
    internal lateinit var password:String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login_fire)

        setupActionBar()

        Name = findViewById(R.id.etName) as EditText
        Password = findViewById(R.id.etPassword) as EditText
        Info = findViewById(R.id.tvInfo) as TextView
        Login = findViewById(R.id.btnLogin) as Button
        Login!!.setOnClickListener(this)
        userRegistration = findViewById(R.id.tvRegister) as TextView
        forgotPassword = findViewById(R.id.tvForgotPassword) as TextView

        Info!!.setText("Numero de intentos restantes: 5")

        firebaseAuth = FirebaseAuth.getInstance()
        progressDialog = ProgressDialog(this)

        val user = firebaseAuth!!.getCurrentUser()

        if (user != null) {
            //finish();

            startActivity(Intent(this@LoginFire, SecondActivity::class.java))

        }

        Login!!.setOnClickListener(View.OnClickListener {
            //validacion();
            if (validacion()!!) {
                validate(Name!!.getText().toString(), Password!!.getText().toString())
            }
        })

        userRegistration!!.setOnClickListener(View.OnClickListener {
            startActivity(
                Intent(
                    this@LoginFire,
                    RegistrationActivity::class.java
                )
            )
        })

        forgotPassword!!.setOnClickListener(View.OnClickListener {
            startActivity(
                Intent(
                    this@LoginFire,
                    PasswordActivity::class.java
                )
            )
        })
    }

    private fun validate(userName: String, userPassword: String) {

        progressDialog?.setMessage("Espere un momento por favor")
        progressDialog?.show()

        firebaseAuth?.signInWithEmailAndPassword(userName, userPassword)?.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                progressDialog?.dismiss()
                //Toast.makeText(MainActivity.this, "Login Successful", Toast.LENGTH_SHORT).show();
                checkEmailVerification()
            } else {
                Toast.makeText(this@LoginFire, "Verifique que los datos estén correctos ", Toast.LENGTH_SHORT).show()
                counter--
                Info?.setText("Numero de intentos restantes: $counter")
                progressDialog?.dismiss()
                if (counter == 0) {
                    Login?.setEnabled(false)
                }
            }
        }
    }

    private fun checkEmailVerification() {
        val firebaseUser = firebaseAuth?.getCurrentUser()
        val emailflag = firebaseUser!!.isEmailVerified()

        startActivity(Intent(this@LoginFire, SecondActivity::class.java))

        //        if(emailflag){
        //            finish();
        //            startActivity(new Intent(MainActivity.this, SecondActivity.class));
        //        }else{
        //            Toast.makeText(this, "Verify your email", Toast.LENGTH_SHORT).show();
        //            firebaseAuth.signOut();
        //        }
    }

    private fun validacion(): Boolean? {

        val result = false

        password = Password?.getText().toString()
        name = Name?.getText().toString()

        /*
        if(name.isEmpty() || password.isEmpty()){
            Toast.makeText(this, "Complete los datos por favor", Toast.LENGTH_SHORT).show();
        }else{
            result = true;
        }
        */
        if (name == "") {
            Name?.setError("Requerido")
        } else if (!Patterns.EMAIL_ADDRESS.matcher(name).matches()) {
            Name?.setError("Por favor, introduce una dirección de correo electrónico válida")
            return false
        } else if (password == "") {
            Password?.setError("Requerido")
        } else if (!PASSWORD_PATTERN.matcher(password).matches()) {
            Password?.setError("La contraseña no sigue los parametros ")
            return false
        } else {
            Password?.setError(null)
            return true
        }

        return result
    }

    private fun validatePassword(): Boolean {

        if (password.isEmpty()) {
            Password?.setError("El campo no puede estar vacío")
            return false
        } else if (!PASSWORD_PATTERN.matcher(password).matches()) {
            Password?.setError("Password too weak")
            return false
        } else {
            Password?.setError(null)
            return true
        }
    }

    override fun onClick(v: View) {

    }

    private fun validateEmail(): Boolean {

        if (name.isEmpty()) {
            Name?.setError("El campo no puede estar vacío")
            return false
        } else if (!Patterns.EMAIL_ADDRESS.matcher(name).matches()) {
            Name?.setError("Por favor, introduce una dirección de correo electrónico válida")
            return false
        } else {
            Name?.setError(null)
            return true
        }
    }

    private fun setupActionBar() {
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setTitle("Login a Firebase")
        }
    }
}
